BTRequestStream: generic handler & exception catching

jagerman opened the pull request:
This adds the ability to construct a BTRequestStream with a generic command handler so that if you have a non-trivial set of commands you don't have to incur a potentially large instantiation cost of constructing a bunch of lambda for `register_command` calls that may never be used.  Instead you can now provide a single handler that gets called for any incoming command.

This also adds the ability to throw a new `no_such_endpoint` exception so that such a handler can signal that an endpoint doesn't exist.

Since this necessitated adding an exception handler anyway, this also catches generic exceptions and wraps them with a generic error-producing response function.  Previously an uncaught exception throw from a handler would crash the program.

jagerman committed:
BTRequestStream: generic handler & exception catching

This adds the ability to construct a BTRequestStream with a generic
command handler so that if you have a non-trivial set of commands you
don't have to incur a potentially large instantiation cost of
constructing a bunch of lambda for `register_command` calls that may
never be used.  Instead you can now provide a single handler that gets
called for any incoming command.

This also adds the ability to throw a new `no_such_endpoint` exception
so that such a handler can signal that an endpoint doesn't exist.

Since this necessitated adding an exception handler anyway, this also
catches generic exceptions and wraps them with a generic error-producing
response function.  Previously an uncaught exception throw from a
handler would crash the program.

dr7ana commented on include/quic/btstream.hpp:
```diff
@@ -224,19 +239,33 @@ namespace oxen::quic
+        // Optional constructor argument: generic request handler.  If set, this is invoked for all
+        // incoming requests to endpoints that do not have a `register_command`.  (And so if
+        // `register_command` is not used at all, this can handle all endpoint).
+        void handle_bp_opt(std::function<void(oxen::quic::message m)> request_handler)
```
Do you think it would be better to use a member function `register_generic_handler`, similar to how we use `register_command` for each specific endpoint?

I don't feel too strongly either way, but the explicit-ness of the method feels like a convincing idea to me. 

If you didn't want to create a new method, we could even change the parameters of  `register_command` like:
```c++
void BTRequestStream::register_command(std::optional<std::string> ep, std::function<void(message)> func)
{
    endpoint.call(
        [this, ep = std::move(ep), func = std::move(func)]() mutable { 
            if (ep)
                func_map[std::move(ep)] = std::move(func);
            else
                generic_handler = std::move(func);
        });
}
```
So setting the generic handler means passing `std::nullopt` to the first argument.

jagerman replied:
I suppose it could have a method for it as well, but it is pretty simple to be able to just slap it into the constructor.

jagerman replied:
There's a second reason for it being useful in the constructor: we don't have to worry about thread synchronization during construction and so can just set it directly, where the method version (which I'm also adding) has to worry about this by deferring the assignment to an event loop call.

dr7ana commented on tests/002-send-receive.cpp:
```diff
@@ -628,4 +628,64 @@ namespace oxen::quic::test
+        Address server_local{};
+        Address client_local{};
+
+        std::mutex mut;
```
Unused and triggered a fail in CI [here](https://ci.oxen.rocks/oxen-io/oxen-libquic/503/5/3)

jagerman committed:
Add `register_command_fallback`, misc fixes

This allows setting (or changing, or clearing) the fallback handler
after construction, alongside other `register_command` calls.

- lowers log levels on bad commands (warning) or exceptions (error)
  instead of criticals.
- removes debug criticals from test.

jagerman committed:
Make response functions `const`

There's nothing in them requiring non-const, and it's not that uncommon
to have a const reference for instance when working with a lambda
capture).

jagerman committed:
Rename register_command to register_handler

"command" is a bit of a misnomer from an early idea where we were
copying oxenmq's distinction of "requests" and "commands", but
BTRequestStream no longer has that distinction (whether to send/accept a
reply is application-level protocol logic).

This renames it to the better name `register_handler` (and renames
`register_command_fallback` to `register_generic_handler` to match).
